File: nodejs/src/stateMachine.ts
import { ActionStatus } from "./types.js";

const TERMINAL = new Set<string>([
  ActionStatus.BLOCKED,
  ActionStatus.DENIED,
  ActionStatus.EXPIRED,
  ActionStatus.SUCCEEDED,
  ActionStatus.FAILED,
  ActionStatus.CANCELLED,
]);

const ALLOWED = new Set<string>(
  [
    [ActionStatus.REQUESTED, ActionStatus.EVALUATED],
    [ActionStatus.REQUESTED, ActionStatus.BLOCKED],
    [ActionStatus.REQUESTED, ActionStatus.FAILED],
    [ActionStatus.EVALUATED, ActionStatus.BLOCKED],
    [ActionStatus.EVALUATED, ActionStatus.WAITING_APPROVAL],
    [ActionStatus.EVALUATED, ActionStatus.EXECUTING],
    [ActionStatus.EVALUATED, ActionStatus.FAILED],
    [ActionStatus.WAITING_APPROVAL, ActionStatus.APPROVED],
    [ActionStatus.WAITING_APPROVAL, ActionStatus.DENIED],
    [ActionStatus.WAITING_APPROVAL, ActionStatus.EXPIRED],
    [ActionStatus.WAITING_APPROVAL, ActionStatus.CANCELLED],
    [ActionStatus.APPROVED, ActionStatus.EXECUTING],
    [ActionStatus.APPROVED, ActionStatus.CANCELLED],
    [ActionStatus.EXECUTING, ActionStatus.SUCCEEDED],
    [ActionStatus.EXECUTING, ActionStatus.FAILED],
    [ActionStatus.EXECUTING, ActionStatus.CANCELLED],
  ].map(([a, b]) => `${a}->${b}`)
);

export class SqubeInvalidStateTransitionError extends Error {
  constructor(
    readonly fromStatus: string,
    readonly toStatus: string,
    readonly executionId?: string
  ) {
    super(
      `Invalid execution transition ${fromStatus} -> ${toStatus}` +
        (executionId ? ` (execution_id=${executionId})` : "")
    );
    this.name = "SqubeInvalidStateTransitionError";
  }
}

export function assertTransition(
  fromStatus: string,
  toStatus: string,
  executionId?: string
): void {
  if (fromStatus === toStatus) {
    throw new SqubeInvalidStateTransitionError(fromStatus, toStatus, executionId);
  }
  if (TERMINAL.has(fromStatus)) {
    throw new SqubeInvalidStateTransitionError(fromStatus, toStatus, executionId);
  }
  const key = `${fromStatus}->${toStatus}`;
  if (!ALLOWED.has(key)) {
    throw new SqubeInvalidStateTransitionError(fromStatus, toStatus, executionId);
  }
}

File: rust/src/state_machine.rs
use crate::error::SqubeGuardError;
use crate::models::ActionStatus;

fn terminal(status: ActionStatus) -> bool {
    matches!(
        status,
        ActionStatus::Blocked
            | ActionStatus::Denied
            | ActionStatus::Expired
            | ActionStatus::Succeeded
            | ActionStatus::Failed
            | ActionStatus::Cancelled
    )
}

fn allowed(from: ActionStatus, to: ActionStatus) -> bool {
    matches!(
        (from, to),
        (ActionStatus::Requested, ActionStatus::Evaluated)
            | (ActionStatus::Requested, ActionStatus::Blocked)
            | (ActionStatus::Requested, ActionStatus::Failed)
            | (ActionStatus::Evaluated, ActionStatus::Blocked)
            | (ActionStatus::Evaluated, ActionStatus::WaitingApproval)
            | (ActionStatus::Evaluated, ActionStatus::Executing)
            | (ActionStatus::Evaluated, ActionStatus::Failed)
            | (ActionStatus::WaitingApproval, ActionStatus::Approved)
            | (ActionStatus::WaitingApproval, ActionStatus::Denied)
            | (ActionStatus::WaitingApproval, ActionStatus::Expired)
            | (ActionStatus::WaitingApproval, ActionStatus::Cancelled)
            | (ActionStatus::Approved, ActionStatus::Executing)
            | (ActionStatus::Approved, ActionStatus::Cancelled)
            | (ActionStatus::Executing, ActionStatus::Succeeded)
            | (ActionStatus::Executing, ActionStatus::Failed)
            | (ActionStatus::Executing, ActionStatus::Cancelled)
    )
}

pub fn assert_transition(from: ActionStatus, to: ActionStatus) -> Result<(), SqubeGuardError> {
    if from == to {
        return Err(SqubeGuardError(format!(
            "Invalid execution transition {} -> {}",
            from.as_str(),
            to.as_str()
        )));
    }
    if terminal(from) {
        return Err(SqubeGuardError(format!(
            "Invalid execution transition {} -> {}",
            from.as_str(),
            to.as_str()
        )));
    }
    if !allowed(from, to) {
        return Err(SqubeGuardError(format!(
            "Invalid execution transition {} -> {}",
            from.as_str(),
            to.as_str()
        )));
    }
    Ok(())
}
